### ISSUE TITLE
Html from Outlook 2016 not being converted to Markdown

### ISSUE THREAD

**Issue by janis-veinbergs** [mysticmind/reversemarkdown-net#61]
So I'v got HTML created by Outlook 2016 with body text: `Test content` and I'v got some wrong outputs from ReverseMarkdown. I'v managed to strip down html to some minimal repro:
```
<html>
<head>
<style><!-- some comment --></style><!-- another comment -->
</head>
<body>
<div>
<p>Test content</p>
</div>
</body>
</html>
```

- If config `RemoveComments = true` and `UnknownTags = Config.UnknownTagsOption.Drop`, then output is empty string, but I would expect `Test content`
- If config `RemoveComments = true` and `UnknownTags = Config.UnknownTagsOption.Bypass`, then output is empty string, but I would expect `Test content`
- If config `RemoveComments = false` and `UnknownTags = Config.UnknownTagsOption.Bypass`, then output is missing `another comment`:
```html
<!-- some comment -->
Test content

```

If I remove `<!-- another comment -->`, then my tests pass. 

The HtmlAgilityPack DOM seems OK to me: [DOM Head node](https://i.imgur.com/0IiTTP3.png) and [DOM P Node](https://i.imgur.com/mqB6Dol.png)

I'm currently investigating the issue deeper.

Raw html from Outlook:
```html
<html xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:w="urn:schemas-microsoft-com:office:word" xmlns:m="http://schemas.microsoft.com/office/2004/12/omml" xmlns="http://www.w3.org/TR/REC-html40">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="Generator" content="Microsoft Word 15 (filtered medium)">
<style><!--
/* Font Definitions */
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
/* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;
	mso-fareast-language:EN-US;}
a:link, span.MsoHyperlink
	{mso-style-priority:99;
	color:#0563C1;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{mso-style-priority:99;
	color:#954F72;
	text-decoration:underline;}
span.EmailStyle17
	{mso-style-type:personal-compose;
	font-family:"Calibri",sans-serif;
	color:windowtext;}
.MsoChpDefault
	{mso-style-type:export-only;
	font-family:"Calibri",sans-serif;
	mso-fareast-language:EN-US;}
@page WordSection1
	{size:612.0pt 792.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;}
div.WordSection1
	{page:WordSection1;}
--></style><!--[if gte mso 9]><xml>
<o:shapedefaults v:ext="edit" spidmax="1026" />
</xml><![endif]--><!--[if gte mso 9]><xml>
<o:shapelayout v:ext="edit">
<o:idmap v:ext="edit" data="1" />
</o:shapelayout></xml><![endif]-->
</head>
<body lang="LV" link="#0563C1" vlink="#954F72">
<div class="WordSection1">
<p class="MsoNormal">Test content<o:p></o:p></p>
</div>
</body>
</html>
```

Test cases used for debugging
```csharp
        [Fact]
        public void WhenOutlook2016Html_WithRemoveCommentsAndDropUnknownTags_ThenConvertToMarkdown()
        {
            // note that the string also has a tab space
            string html = @"<html>
<head>
<style><!-- some comment --></style><!-- another comment -->
</head>
<body>
<div>
<p>Test content</p>
</div>
</body>
</html>";
            string expected = $"{Environment.NewLine}Test content{Environment.NewLine}";

            CheckConversion(html, expected, new Config() {
                RemoveComments = true,
                UnknownTags = Config.UnknownTagsOption.Drop,
            });
        }


        [Fact]
        public void WhenOutlook2016Html_WithRemoveCommentsAndBypassUnknownTags_ThenConvertToMarkdown() {
            // note that the string also has a tab space
            string html = @"<html>
<head>
<style><!-- some comment --></style><!-- another comment -->
</head>
<body>
<div>
<p>Test content</p>
</div>
</body>
</html>";
            string expected = $"{Environment.NewLine}Test content{Environment.NewLine}";

            CheckConversion(html, expected, new Config() {
                RemoveComments = true,
                UnknownTags = Config.UnknownTagsOption.Bypass,
            });
        }

        [Fact]
        public void WhenOutlook2016Html_WithDoNotRemoveCommentsAndDropUnknownTags_ThenConvertToMarkdown() {
            // note that the string also has a tab space
            string html = @"<html>
<head>
<style><!-- some comment --></style><!-- another comment -->
</head>
<body>
<div>
<p>Test content</p>
</div>
</body>
</html>";
            string expected = $"<!-- some comment -->{Environment.NewLine}Test content{Environment.NewLine}";

            CheckConversion(html, expected, new Config() {
                RemoveComments = false,
                UnknownTags = Config.UnknownTagsOption.Bypass,
            });
        }
```

**Comment by janis-veinbergs** [comment]
So given `RemoveComments = true` and after [Cleaner.RemoveComments](https://github.com/mysticmind/reversemarkdown-net/blob/2a29c36c5ade6534fb5fdb99f1aea53b8a4ac5c2/src/ReverseMarkdown/Cleaner.cs#L17) the HTML is:
```html
<html>
<head>
<style>
</head>
<body>
<div>
<p>Test content</p>
</div>
</body>
</html>
```

The Regex strips the `</style>` closing tag. There is a [famous answer on stackoverflow](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags) on why Regex shouldn't be used for HTML parsing. 

The other question in my mind - shouldn't HtmlAgilityPack be vigilant enought to handle unclosed tags?

That led me to even shorter testcase:
```csharp
        [Fact]
        public void WhenUnclosedStyleTag_WithBypassUnknownTags_ThenConvertToMarkdown() {
            // note that the string also has a tab space
            string html = @"<html><head><style></head><body><p>Test content</p></body></html>";
            string expected = $"{Environment.NewLine}Test content{Environment.NewLine}";

            CheckConversion(html, expected, new Config() {
                UnknownTags = Config.UnknownTagsOption.Bypass
            });
        }
```
Result:
```
Expected: 
Test content

Actual:   
```

So the unclosed style tag is the culprit here. From one hand, I think RemoveComments should be fixed: either regex fixed or when traversing DOM, comments should be ignored when `RemoveComments = true`.

The regex fix is quite simple, actually. `<!--(\n|.)*-->` should be changed to  `<!--(\n|.)*?-->` (Will stop matching as soon as first `-->` encountered)

On the other side, comments just brought to the bug which is not directly related to comments but unclosed tag. And that should be handled somehow too. Perhaps there are flags that can be passed to HtmlAgilityPack to change its behavior. Maybe it is related to overlapping tags? https://stackoverflow.com/questions/4218847/htmlagilitypack-does-form-close-itself-for-some-reason
I'll try to investigate.

**Comment by janis-veinbergs** [comment]
Regarding unclosed style tag. Yeah, that kind of breaks HtmlAgilityPack.

When specifying `HtmlNode.ElementsFlags["style"] = HtmlElementFlag.Empty;` in [Converter constructor](https://github.com/mysticmind/reversemarkdown-net/blob/9c71437b190f7f16650c8be7e1e3e2a914a479eb/src/ReverseMarkdown/Converter.cs#L19), the issue is resolved. However the built-in flag is `HtmlElementFlag.CData`. Specifying `HtmlElementFlag.CData | HtmlElementFlag.Empty` still has parsing issues.

I wonder if this should be asked in HtmlAgilityPack issue tracker. 

As of this case, i'll do pull request for updated regex for comments removal. That fixes incoming emails from Outlook.

However as of other case, when [CData tags](https://github.com/pinscript/HtmlAgilityPack/blob/d36ae585beb9a2d879c717aab7e91ecb2cc8d4d4/Source/HtmlAgilityPack/HtmlNode.cs#L81): `style`, `script`, `noxhtml (?)` may be unclosed - remains to be addressed.

P.S. Project uses outdated HtmlAgilityPack. I'v tried updating `1.11.1` to `1.11.18`, the issue persists.

**Comment by janis-veinbergs** [comment]
The linked issue https://github.com/zzzprojects/html-agility-pack/issues/372 shows that closing style/script is a MUST and even browser breaks under that case.

So I think it is fair then if HtmlAgilityPack doesn't try to be better/more graceful than browsers are (which usually handle broken HTML well). 

So I think this issue should be considered closed if maintainer accepts the fix in pull request #62 regarding comments.

**Comment by mysticmind** [comment]
Fixed via #62

**Comment by mysticmind** [comment]
Available in v3.11.0 published few mins ago
